File: src/main/java/com/rs/game/content/minigames/ectofuntus/Ectofuntus.java
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//  Copyright (C) 2021 Trenton Kress
//  This file is part of project: Darkan
//
package com.rs.game.content.minigames.ectofuntus;

import com.rs.cache.loaders.ItemDefinitions;
import com.rs.cache.loaders.ObjectDefinitions;
import com.rs.game.content.skills.agility.Agility;
import com.rs.game.content.skills.magic.Magic;
import com.rs.game.content.skills.prayer.Burying;
import com.rs.game.model.entity.player.Player;
import com.rs.lib.Constants;
import com.rs.lib.game.Animation;
import com.rs.lib.game.Item;
import com.rs.lib.game.Tile;
import com.rs.plugin.annotations.PluginEventHandler;
import com.rs.plugin.handlers.ObjectClickHandler;

import java.util.HashMap;
import java.util.Map;

@PluginEventHandler
public class Ectofuntus {

	public static final int EMPTY_POT = 1931;
	public static final int EMPTY_BUCKET = 1925;
	public static final int ECTO_TOKEN = 4278;
	public static final int BUCKET_OF_SLIME = 4286;
	public static final int ECTOPHIAL = 4251;

	public static enum BoneMeal {
		BONES(526, 4255),
		BAT_BONES(530, 4256),
		BIG_BONES(532, 4257),
		BABY_DRAGON_BONES(534, 4260),
		DRAGON_BONES(536, 4261),
		DAGANNOTH_BONES(6729, 6728),
		WYVERN_BONES(6812, 6810),
		OURG_BONES(4834, 4855),
		FROST_BONES(18832, 18834),
		IMPIOUS_ASHES(20264, 20264),
		ACCURSED_ASHES(20266, 20266),
		INFERNAL_ASHES(20268, 20268);

		private int boneId;
		private int boneMealId;

		private static Map<Integer, BoneMeal> bonemeals = new HashMap<>();
		private static Map<Integer, BoneMeal> bones = new HashMap<>();

		public static BoneMeal forBoneId(int itemId) {
			return bonemeals.get(itemId);
		}

		public static BoneMeal forMealId(int itemId) {
			return bones.get(itemId);
		}

		static {
			for (final BoneMeal bonemeal : BoneMeal.values())
				bonemeals.put(bonemeal.boneId, bonemeal);
			for (final BoneMeal bonemeal : BoneMeal.values())
				bones.put(bonemeal.boneMealId, bonemeal);
		}

		private BoneMeal(int boneId, int boneMealId) {
			this.boneId = boneId;
			this.boneMealId = boneMealId;
		}

		public int getBoneId() {
			return boneId;
		}

		public int getBoneMealId() {
			return boneMealId;
		}
	}

	public static final void sendEctophialTeleport(Player player, Tile tile) {
		Magic.sendTeleportSpell(player, 8939, 8941, 1678, 1679, 0, 0, tile, 3, true, Magic.MAGIC_TELEPORT, null);
	}

	public static ObjectClickHandler handleEntrance = new ObjectClickHandler(new Object[] { 5268 }, e -> e.getPlayer().useLadder(Tile.of(3669, 9888, 3)));
	public static ObjectClickHandler handleExit = new ObjectClickHandler(new Object[] { 5264 }, e -> e.getPlayer().useLadder(Tile.of(3654, 3519, 0)));

	public static ObjectClickHandler handleShortcuts = new ObjectClickHandler(new Object[] { 9307, 9308 }, e -> {
		switch (e.getObjectId()) {
		case 9307:
			if (!Agility.hasLevel(e.getPlayer(), 53))
				return;
			e.getPlayer().useLadder(Tile.of(3670, 9888, 3));
			return;

		case 9308:
			if (!Agility.hasLevel(e.getPlayer(), 53))
				return;
			e.getPlayer().useLadder(Tile.of(3671, 9888, 2));
			return;
		}
	});

	public static ObjectClickHandler handleStairs = new ObjectClickHandler(new Object[] { 5262, 5263 }, e -> {
		switch (e.getObjectId()) {
		case 5262:
			if (e.getPlayer().getPlane() == 2)
				e.getPlayer().setNextTile(Tile.of(3692, 9888, 3));
			if (e.getPlayer().getPlane() == 1)
				e.getPlayer().setNextTile(Tile.of(3671, 9888, 2));
			if (e.getPlayer().getPlane() == 0)
				e.getPlayer().setNextTile(Tile.of(3687, 9888, 1));
			return;

		case 5263:
			if (e.getPlayer().getPlane() == 3)
				e.getPlayer().setNextTile(Tile.of(3688, 9888, 2));
			if (e.getPlayer().getPlane() == 2)
				e.getPlayer().setNextTile(Tile.of(3675, 9887, 1));
			if (e.getPlayer().getPlane() == 1)
				e.getPlayer().setNextTile(Tile.of(3683, 9888, 0));
			return;
		}
	});

	public static ObjectClickHandler handleWorship = new ObjectClickHandler(new Object[] { 5282 }, e -> {
		if (!e.getPlayer().getInventory().containsItem(BUCKET_OF_SLIME, 1)) {
			e.getPlayer().sendMessage("You need a bucket of slime before you can worship the ectofuntus.");
			return;
		}
		for (Item item : e.getPlayer().getInventory().getItems().array()) {
			if (item == null)
				continue;
			BoneMeal bone = BoneMeal.forMealId(item.getId());
			if (bone != null) {
				Burying.Bone boneData = Burying.Bone.forId(bone.getBoneId());
				if (boneData == null) {
					e.getPlayer().sendMessage("Error bone not added.. Please post the bone you tried to add on the forums.");
					return;
				}
				e.getPlayer().incrementCount(ItemDefinitions.getDefs(bone.getBoneId()).getName()+" offered at ectofuntus");
				e.getPlayer().setNextAnimation(new Animation(1651));
				e.getPlayer().getInventory().deleteItem(bone.getBoneMealId(), 1);
				if (!bone.name().contains("ASHES"))
					e.getPlayer().getInventory().addItem(EMPTY_POT, 1);
				e.getPlayer().getInventory().deleteItem(BUCKET_OF_SLIME, 1);
				e.getPlayer().getInventory().addItem(EMPTY_BUCKET, 1);
				e.getPlayer().getSkills().addXp(Constants.PRAYER, boneData.getExperience() * 4);
				e.getPlayer().unclaimedEctoTokens += 5;
				return;
			}
		}
	});

	public static ObjectClickHandler handleGrinder = new ObjectClickHandler(new Object[] { 11163 }, e -> grinder(e.getPlayer()));
	public static ObjectClickHandler handleBin = new ObjectClickHandler(new Object[] { 11164 }, e -> bin(e.getPlayer()));

	public static boolean handleItemOnObject(Player player, int itemId, int objectId) {
		ObjectDefinitions objectDefs = ObjectDefinitions.getDefs(objectId);
		ItemDefinitions itemDefs = ItemDefinitions.getDefs(itemId);

		if (itemId == EMPTY_BUCKET && objectDefs.getName().equals("Pool of Slime")) {
			player.getActionManager().setAction(new SlimeBucketFill());
			return true;
		}

		if (itemDefs.getName().toLowerCase().contains("bone") && objectId == 11162) {
			player.getActionManager().setAction(new EctoBones(itemId));
			return true;
		}
		return false;
	}

	public static boolean hopper(Player player, int itemId) {
		if (player.boneType != -1 && player.boneType != 0) {
			player.sendMessage("You already have some bones in the hopper.");
			return false;
		}
		if (!player.getInventory().containsItem(itemId, 1))
			return false;
		BoneMeal meal = BoneMeal.forBoneId(itemId);
		if (meal != null) {
			player.boneType = meal.getBoneId();
			player.sendMessage("You put the bones in the hopper.", true);
			player.setNextAnimation(new Animation(1649));
			player.getInventory().deleteItem(meal.getBoneId(), 1);
			return true;
		}
		player.boneType = -1;
		return false;
	}

	public static boolean grinder(Player player) {
		if (player.boneType != -1 && !player.bonesGrinded) {
			player.sendMessage("You turn the grinder, some crushed bones fall into the bin.", true);
			player.setNextAnimation(new Animation(1648));
			player.bonesGrinded = true;
			return true;
		}
		player.setNextAnimation(new Animation(1648));
		return false;
	}

	public static boolean bin(Player player) {
		if (player.boneType == -1) {
			player.sendMessage("You need to put some bones in the hopper and grind them first.");
			return false;
		}
		if (!player.bonesGrinded) {
			player.sendMessage("You need to grind the bones by turning the grinder first.");
			return false;
		}
		if (player.boneType != -1 && player.bonesGrinded) {
			BoneMeal meal = BoneMeal.forBoneId(player.boneType);
			if (meal != null) {
				player.sendMessage("You fill an empty pot with bones.", true);
				player.setNextAnimation(new Animation(1650));
				player.getInventory().deleteItem(EMPTY_POT, 1);
				player.getInventory().addItem(meal.getBoneMealId(), 1);
				player.boneType = -1;
				player.bonesGrinded = false;
				return true;
			}
			player.boneType = -1;
		}
		return false;
	}
}

File: src/main/java/com/rs/game/content/minigames/ectofuntus/Ectofuntus.kt
package com.rs.game.content.minigames.ectofuntus

import com.rs.cache.loaders.ItemDefinitions
import com.rs.cache.loaders.ObjectDefinitions
import com.rs.engine.quest.Quest
import com.rs.game.World
import com.rs.game.content.skills.agility.Agility
import com.rs.game.content.skills.magic.Magic
import com.rs.game.content.skills.magic.TeleType
import com.rs.game.content.skills.prayer.Burying
import com.rs.game.model.entity.async.schedule
import com.rs.game.model.entity.player.Player
import com.rs.lib.Constants
import com.rs.lib.game.Tile
import com.rs.plugin.annotations.ServerStartupEvent
import com.rs.plugin.kts.onItemClick
import com.rs.plugin.kts.onItemOnObject
import com.rs.plugin.kts.onObjectClick

class Ectofuntus {

    companion object {
        const val EMPTY_POT = 1931
        const val EMPTY_BUCKET = 1925
        const val ECTO_TOKEN = 4278
        const val BUCKET_OF_SLIME = 4286
        const val FULL_ECTOPHIAL = 4251
        const val EMPTY_ECTOPHIAL = 4252
        const val ECTOFUNTUS = 5282

        @JvmStatic
        fun handleItemOnObject(player: Player, itemId: Int, objectId: Int): Boolean {
            val objectDefs = ObjectDefinitions.getDefs(objectId)
            val itemDefs = ItemDefinitions.getDefs(itemId)
            if (itemId == EMPTY_BUCKET && objectDefs.name == "Pool of Slime") {
                player.actionManager.setAction(SlimeBucketFill())
                return true
            }
            if (itemDefs.name.lowercase().contains("bone") && objectId == 11162) {
                player.actionManager.setAction(EctoBones(itemId))
                return true
            }
            return false
        }
    }

}

@ServerStartupEvent
fun mapEctofuntusInteractions() {

    onObjectClick(5268) { (player) -> player.useLadder(Tile.of(3669, 9888, 3)) } // Entrance to Ectofuntus Pool
    onObjectClick(5264) { (player) -> player.useLadder(Tile.of(3654, 3519, 0)) } // Exit from Ectofuntus Pool

    onObjectClick(9307, 9308) { (player, obj) ->  // Ectofuntus Pool Shortcuts
        when (obj.id) {
            9307 -> {
                if (!Agility.hasLevel(player, 53)) return@onObjectClick
                player.useLadder(Tile.of(3670, 9888, 3))
            }

            9308 -> {
                if (!Agility.hasLevel(player, 53)) return@onObjectClick
                player.useLadder(Tile.of(3671, 9888, 2))
            }
        }
    }

    onObjectClick(5262, 5263) { (player, obj) ->  // Ectofuntus Pool Stairs
        when (obj.id) {
            5262 -> {
                when (player.plane) {
                    2 -> player.tele(Tile.of(3692, 9888, 3))
                    1 -> player.tele(Tile.of(3671, 9888, 2))
                    0 -> player.tele(Tile.of(3687, 9888, 1))
                }
            }

            5263 -> {
                when (player.plane) {
                    3 -> player.tele(Tile.of(3688, 9888, 2))
                    2 -> player.tele(Tile.of(3675, 9887, 1))
                    1 -> player.tele(Tile.of(3683, 9888, 0))
                }
            }
        }
    }

    onObjectClick(Ectofuntus.ECTOFUNTUS) { (player) ->  // Ectofuntus Worship
        if (!player.inventory.containsItem(Ectofuntus.BUCKET_OF_SLIME, 1)) {
            player.sendMessage("You need a bucket of slime before you can worship at the ectofuntus.")
            return@onObjectClick
        }

        var hasBonemeal = false
        for (item in player.inventory.items.toArray()) {
            if (item == null) continue

            val bone = BoneMeal.forMealId(item.id)
            if (bone != null) {
                hasBonemeal = true
                break
            }
        }

        if (!hasBonemeal) {
            player.sendMessage("You need some bonemeal in order to worship at the ectofuntus.")
            return@onObjectClick
        }

        for (item in player.inventory.items.toArray()) {
            if (item == null) continue

            val bone = BoneMeal.forMealId(item.id)
            if (bone != null) {
                val boneData = Burying.Bone.forId(bone.boneId)
                if (boneData == null) {
                    player.sendMessage("Error - bone not added... Please submit a Bug Report on Discord with the bone type you tried to add.")
                    return@onObjectClick
                }

                player.incrementCount("${ItemDefinitions.getDefs(bone.boneId).name} offered at ectofuntus")
                player.anim(1651)
                player.inventory.deleteItem(bone.boneMealId, 1)
                player.inventory.deleteItem(Ectofuntus.BUCKET_OF_SLIME, 1)
                player.inventory.addItem(Ectofuntus.EMPTY_POT, 1)
                player.inventory.addItem(Ectofuntus.EMPTY_BUCKET, 1)
                player.skills.addXp(Constants.PRAYER, boneData.experience * 4)
                player.unclaimedEctoTokens += 5
                break
            }
        }
    }

    onObjectClick(37454) { (player, obj) -> player.useStairs(-1, Tile.of(player.x, obj.y + 5, player.plane + 1), 0, 1) } // Ectofuntus Grinder Stairs Up
    onObjectClick(5281) { (player, obj) -> player.useStairs(-1, Tile.of(player.x, obj.y - 4, player.plane - 1), 0, 1) } // Ectofuntus Grinder Stairs Down

    onObjectClick(11163) { (player) -> grinder(player) } // Bone Grinder
    onObjectClick(11164) { (player) -> bin(player) } // Bone Grinder Bin

    onItemClick(Ectofuntus.FULL_ECTOPHIAL) { (player) -> // Empty Ectophial (Teleport)
        if (!player.isQuestComplete(Quest.GHOSTS_AHOY, "to use the ectophial.")) return@onItemClick

        player.schedule {
            player.anim(9609)
            player.spotAnim(1688)
            player.sendMessage("You empty your ectophial on the floor...", true)
            player.lock(3)
            wait(5)
            Magic.sendTeleportSpell(
                player, 8939, 8941, 1678, 1679, 0, 0.0, Tile.of(3659, 3523, 0), 3, true, TeleType.MAGIC,
                Runnable {
                    player.lock()
                    player.soundEffect(4580, true)
                },
                Runnable {
                    player.unlock()
                    player.sendMessage("Your ectophial magically fills itself from the nearby ectofuntus.", true)
                    /*val obj = World.getObject(Tile.of(3658, 3518, 0))
                    if (obj != null) {
                        player.inventory.replace(Ectofuntus.FULL_ECTOPHIAL, Ectofuntus.EMPTY_ECTOPHIAL)
                        player.walkToAndExecute(World.findClosestAdjacentFreeTile(Tile.of(3659, 3521, 0), 1)) {
                            player.schedule {
                                wait(1)
                                player.faceTile(obj.tile)
                                wait(1)
                                player.anim(833)
                                player.inventory.replace(Ectofuntus.EMPTY_ECTOPHIAL, Ectofuntus.FULL_ECTOPHIAL)
                                player.sendMessage("You refill your ectophial from the ectofuntus.")
                            }
                        }
                    }*/
                }
            )
        }
    }

    onItemOnObject(arrayOf(Ectofuntus.ECTOFUNTUS), arrayOf(Ectofuntus.EMPTY_ECTOPHIAL, Ectofuntus.FULL_ECTOPHIAL)) { e -> // Refill ectophial
        when (e.item.id) {
            Ectofuntus.EMPTY_ECTOPHIAL -> {
                e.player.anim(833)
                e.player.inventory.replace(Ectofuntus.EMPTY_ECTOPHIAL, Ectofuntus.FULL_ECTOPHIAL);
                e.player.sendMessage("You refill your ectophial from the ectofuntus.");
            }
            Ectofuntus.FULL_ECTOPHIAL -> e.player.sendMessage("Your ectophial is already full.")
        }
    }

}

fun hopper(player: Player, itemId: Int): Boolean {
    if (player.boneType != -1 && player.boneType != 0) {
        player.sendMessage("You already have some bones in the hopper.")
        return false
    }
    if (!player.inventory.containsItem(itemId, 1)) {
        return false
    }
    val meal = BoneMeal.forBoneId(itemId)
    if (meal != null) {
        player.boneType = meal.boneId
        player.sendMessage("You put the bones in the hopper.", true)
        player.anim(1649)
        player.inventory.deleteItem(meal.boneId, 1)
        return true
    }
    player.boneType = -1
    return false
}

fun grinder(player: Player): Boolean {
    return if (player.boneType != -1 && !player.bonesGrinded) {
        player.sendMessage("You turn the grinder, some crushed bones fall into the bin.", true)
        player.anim(1648)
        player.bonesGrinded = true
        true
    } else {
        player.anim(1648)
        false
    }
}

fun bin(player: Player): Boolean {
    if (player.boneType == -1) {
        player.sendMessage("You need to put some bones in the hopper and grind them first.")
        return false
    }
    if (!player.bonesGrinded) {
        player.sendMessage("You need to grind the bones by turning the grinder first.")
        return false
    }
    if (player.boneType != -1 && player.bonesGrinded) {
        val meal = BoneMeal.forBoneId(player.boneType)
        if (meal != null) {
            player.sendMessage("You fill an empty pot with bones.", true)
            player.anim(1650)
            player.inventory.deleteItem(Ectofuntus.EMPTY_POT, 1)
            player.inventory.addItem(meal.boneMealId, 1)
            player.boneType = -1
            player.bonesGrinded = false
            return true
        }
        player.boneType = -1
    }
    return false
}

enum class BoneMeal(val boneId: Int, val boneMealId: Int) {
    BONES(526, 4255),
    BAT_BONES(530, 4256),
    BIG_BONES(532, 4257),
    BABY_DRAGON_BONES(534, 4260),
    DRAGON_BONES(536, 4261),
    DAGANNOTH_BONES(6729, 6728),
    WYVERN_BONES(6812, 6810),
    OURG_BONES(4834, 4855),
    OURG_BONES_2(14793, 4855),
    FROST_BONES(18832, 18834),
    IMPIOUS_ASHES(20264, 20264),
    ACCURSED_ASHES(20266, 20266),
    INFERNAL_ASHES(20268, 20268);

    companion object {
        private val bonemeals = BoneMeal.entries.associateBy { it.boneId }
        private val bones = BoneMeal.entries.associateBy { it.boneMealId }

        fun forBoneId(itemId: Int): BoneMeal? = bonemeals[itemId]
        fun forMealId(itemId: Int): BoneMeal? = bones[itemId]
    }
}
